### PR TITLE
Cache hasAvatar info in sessionStorage to avoid flicker

### PR DESCRIPTION
**Pull request by PVince81**
This is a crude POC that shows that when we save the "does this user has an avatar call", performance improves.

Possible solution for https://github.com/nextcloud/spreed/issues/4284

Maybe this should be stored in a store or something, but as I understand, Vue components in the library don't really have a store available.

Another approach would be to supply an optional property "hasAvatar" and the caching could be done from the outside in situations where the caller thinks it makes sense.

### CODE REVIEW

**Comment by PVince81** [comment]
Any thoughts on this approach ? @ma12-co @nickvergessen @skjnldsv 

If we are fine caching this information, where would be a better place ? Ideally that info should be available to other components as well like the VideoBackground in the Talk app which uses another separate query to get the avatar to put into the background in blur mode.

File: src/components/Avatar/Avatar.vue [hunk 1]
```diff
@@ -466,17 +466,35 @@ export default {
 				urlGenerator(this.user, this.size * 4) + ' 4x',
 			].join(', ')
 
+			// skip loading
+			if (!window.userHasAvatar) {
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply src/components/Avatar/Avatar.vue:470]: Yes, we are cleaning the window over the last nc releases, we should try our best not to pollute the window namespace anymore :)

[the remark this answers is not in the document]
Comment (PVince81) [reply src/components/Avatar/Avatar.vue:470]: if we were to store information available for all apps, where would that be ?

in the past it used to be in the OC namespace

[the remark this answers is not in the document]
Comment (skjnldsv) [reply src/components/Avatar/Avatar.vue:470]: browserstorage is good enough I think, or localstorage if you want :shrug:

[the remark this answers is not in the document]
Comment (PVince81) [reply src/components/Avatar/Avatar.vue:470]: ok sounds good.

will likely also need some form of expiration, but I'm not sure if worth it as the use case would be that someone deletes their avatar and do not set a new one, which is rather rare

[the remark this answers is not in the document]
Comment (PVince81) [reply src/components/Avatar/Avatar.vue:470]: I think I'll go with sessionStorage. It only lives within the current page and should be enough for the use case: preventing flickering within a call session.

Then it's easier to tell people to force refresh the page than clear cache whenever they noticed that an avatar of a collegue did not get updated.

**Comment by PVince81** [comment]
Ready to review.

Note: I went for the format "userHasAvatar-" + userId instead of a single object to avoid repeated serialization/deserialization (also semi-concurrently)

**Comment by PVince81** [comment]
Weird, I just checked the browser console and somehow nothing appears under sessionStorage.

and also somehow I didn't see flickering any more. I don't understand how it can be working.

investigating...

**Comment by PVince81** [comment]
ok got it... `null` is returned by the sessionStorage instead of `undefined`, which the current logic treats as always `true`... so no flickering...

File: src/components/Avatar/Avatar.vue [hunk 2]
```diff
@@ -100,6 +100,14 @@ import Tooltip from '../../directives/Tooltip'
 import usernameToColor from '../../functions/usernameToColor'
 import { userStatus } from '../../mixins'
 
+function getUserHasAvatar(userId) {
+	return window.sessionStorage.getItem('userHasAvatar-' + userId)
```
Comment (juliushaertl) [inline src/components/Avatar/Avatar.vue:104]: We could probably apply a similar naming schema as we discusssed for the event names in https://github.com/nextcloud/nextcloud-event-bus/issues/131#issuecomment-683689107 That way we still can figure out where a value might be coming from when we fill out browser/session storage more over time.

@skjnldsv @ChristophWurst Didn't we have some planned prefix for the components library as well?

Basically someting like `vue-components:user-has-avatar.userID`

Comment (skjnldsv) [reply src/components/Avatar/Avatar.vue]: Nah, components were generic, so `nextcloud:user-has-avatar.userID` iirc :thinking:

**Comment by PVince81** [comment]
works nicely with Chrome, but in Firefox I still observe a flicker sometimes when hitting "stop following", despite cache hit.

let's still move forward with this as it at least helps resolve the annoying flicker on speaker change, which is already a big step forward

**Review by skjnldsv — CHANGES_REQUESTED** [review]
See https://github.com/nextcloud/nextcloud-vue/pull/1457#discussion_r505552471

File: src/components/Avatar/Avatar.vue (hunk 2 above)
Comment (ChristophWurst) [reply src/components/Avatar/Avatar.vue]: https://github.com/nextcloud/nextcloud-browser-storage :thinking:

Comment (PVince81) [reply src/components/Avatar/Avatar.vue]: ah, that's the global thing I was looking for, thanks :-)

I'll adjust

**Comment by PVince81** [comment]
adjusted in https://github.com/nextcloud/nextcloud-vue/pull/1457/commits/7613d44ce026fe88502206279abf130e26e50143

I decided against introduce a service for just this once single occurrence, considering that there are no services in nextcloud-vue. If this becomes commonplace later on we could adjust.

Please also check https://github.com/nextcloud/spreed/pull/4379 (last commit) which uses the same storage location for the video background.

File: src/components/Avatar/Avatar.vue [hunk 3]
```diff
@@ -100,6 +101,20 @@ import Tooltip from '../../directives/Tooltip'
 import usernameToColor from '../../functions/usernameToColor'
 import { userStatus } from '../../mixins'
 
+const browserStorage = getBuilder('nextcloud').persist().build()
```
Comment (ChristophWurst) [inline src/components/Avatar/Avatar.vue:104]: but now it will use `localStorage` and not `sessionStorage` – intended?

Comment (PVince81) [reply src/components/Avatar/Avatar.vue:104]: that's fine. will increase performance for an even longer time then.

**Comment by PVince81** [comment]
good to merge ? :smile:

**Review by ChristophWurst — APPROVED** [review]

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**